File: packages/snow_draw_core/lib/draw/elements/types/arrow/elbow/elbow_heading.dart
/// Cardinal directions used by elbow routing and editing logic.
enum ElbowHeading { right, down, left, up }

extension ElbowHeadingX on ElbowHeading {
  /// Unit step along the X axis for this heading.
  int get dx => switch (this) {
    ElbowHeading.right => 1,
    ElbowHeading.left => -1,
    _ => 0,
  };

  /// Unit step along the Y axis for this heading.
  int get dy => switch (this) {
    ElbowHeading.down => 1,
    ElbowHeading.up => -1,
    _ => 0,
  };

  /// Whether the heading moves along the X axis.
  bool get isHorizontal =>
      this == ElbowHeading.right || this == ElbowHeading.left;

  /// The opposite heading (used for backtracking checks).
  ElbowHeading get opposite => switch (this) {
    ElbowHeading.right => ElbowHeading.left,
    ElbowHeading.left => ElbowHeading.right,
    ElbowHeading.up => ElbowHeading.down,
    ElbowHeading.down => ElbowHeading.up,
  };
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/elbow/elbow_heading.rs
#![allow(dead_code)]

/// Cardinal directions used by elbow routing and editing logic.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ElbowHeading {
    Right,
    Down,
    Left,
    Up,
}

impl ElbowHeading {
    /// Unit step along the X axis for this heading.
    pub const fn dx(self) -> i32 {
        match self {
            Self::Right => 1,
            Self::Down => 0,
            Self::Left => -1,
            Self::Up => 0,
        }
    }

    /// Unit step along the Y axis for this heading.
    pub const fn dy(self) -> i32 {
        match self {
            Self::Right => 0,
            Self::Down => 1,
            Self::Left => 0,
            Self::Up => -1,
        }
    }

    /// Whether the heading moves along the X axis.
    pub const fn is_horizontal(self) -> bool {
        matches!(self, Self::Right | Self::Left)
    }

    /// The opposite heading (used for backtracking checks).
    pub const fn opposite(self) -> Self {
        match self {
            Self::Right => Self::Left,
            Self::Left => Self::Right,
            Self::Up => Self::Down,
            Self::Down => Self::Up,
        }
    }
}
